File: serve.js
#!/usr/bin/env node

/**
 * Simple HTTP server for LM Studio Mobile App
 * Supports single-page app routing for Expo web export
 *
 * Usage: node serve.js
 */

const http = require('http');
const fs = require('fs');
const path = require('path');
const url = require('url');

const PORT = process.env.PORT || 8000;

// 自动检测静态文件目录：
// - 如果当前目录下有 dist，使用 dist（本地开发/导出后的场景）
// - 否则使用当前目录（gh-pages 分支克隆后的场景）
function findServeDir() {
  const distPath = path.join(__dirname, 'dist');
  if (fs.existsSync(distPath) && fs.statSync(distPath).isDirectory()) {
    // 确保 dist 里有 HTML 文件（避免空目录误判）
    const files = fs.readdirSync(distPath);
    if (files.some(f => f.endsWith('.html'))) {
      return distPath;
    }
  }

  // 检查当前目录是否包含构建产物
  const currentFiles = fs.readdirSync(__dirname);
  if (currentFiles.some(f => f.endsWith('.html')) && currentFiles.includes('_expo')) {
    return __dirname;
  }

  return distPath; // 默认返回 dist 路径，后续会报错提示
}

const DIST_DIR = findServeDir();

const mimeTypes = {
  '.html': 'text/html',
  '.js': 'application/javascript',
  '.css': 'text/css',
  '.json': 'application/json',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.gif': 'image/gif',
  '.svg': 'image/svg+xml',
  '.ico': 'image/x-icon',
  '.woff': 'font/woff',
  '.woff2': 'font/woff2',
  '.ttf': 'font/ttf',
  '.eot': 'application/vnd.ms-fontobject',
};

const server = http.createServer((req, res) => {
  const urlPath = url.parse(req.url).pathname;
  let filePath = path.join(DIST_DIR, urlPath);

  // If path is a directory or doesn't have extension, try HTML file
  if (!path.extname(filePath)) {
    const htmlPath = filePath + '.html';
    if (fs.existsSync(htmlPath)) {
      filePath = htmlPath;
    } else if (fs.existsSync(filePath) && fs.statSync(filePath).isDirectory()) {
      filePath = path.join(filePath, 'index.html');
    }
  }

  // Prevent directory traversal
  if (!filePath.startsWith(DIST_DIR)) {
    res.writeHead(403);
    res.end('Forbidden');
    return;
  }

  // Try to serve the file
  fs.readFile(filePath, (err, data) => {
    if (err) {
      if (err.code === 'ENOENT') {
        // Try to serve root index.html for SPA routing
        fs.readFile(path.join(DIST_DIR, 'index.html'), (err, data) => {
          if (err) {
            res.writeHead(404);
            res.end('Not Found');
            return;
          }
          res.writeHead(200, { 'Content-Type': 'text/html' });
          res.end(data);
        });
      } else {
        res.writeHead(500);
        res.end('Server Error');
      }
      return;
    }

    const ext = path.extname(filePath);
    const contentType = mimeTypes[ext] || 'application/octet-stream';

    // Add cache control headers
    res.writeHead(200, {
      'Content-Type': contentType,
      'Cache-Control': 'no-store, no-cache, must-revalidate, max-age=0',
      'Pragma': 'no-cache',
      'Expires': '0'
    });
    res.end(data);

    // Log request
    console.log(`[${new Date().toLocaleTimeString()}] ${req.method} ${req.url} - ${contentType}`);
  });
});

// 检查静态文件目录是否可用
if (!fs.existsSync(DIST_DIR)) {
  console.error('❌ 错误：找不到静态文件目录');
  console.error('');
  console.error('请确认以下情况之一：');
  console.error('  1. 源码目录：运行 npx expo export --platform web --output-dir ./dist');
  console.error('  2. gh-pages 分支：直接在克隆目录运行 node serve.js');
  process.exit(1);
}

const hasHtml = fs.readdirSync(DIST_DIR).some(f => f.endsWith('.html'));
if (!hasHtml) {
  console.error(`❌ 错误：${DIST_DIR} 中没有找到 HTML 文件`);
  console.error('请先运行 npx expo export --platform web --output-dir ./dist');
  process.exit(1);
}

// Start server
server.listen(PORT, () => {
  console.log('='.repeat(60));
  console.log('🚀 LM Studio Mobile App - Web Server');
  console.log('='.repeat(60));
  console.log(`\n📂 静态目录: ${DIST_DIR}`);
  console.log(`✅ 服务运行中: http://localhost:${PORT}\n`);
  console.log('🎯 在浏览器打开:');
  console.log(`   http://localhost:${PORT}`);
  console.log(`   http://127.0.0.1:${PORT}\n`);
  console.log('⚠️  请确保 LM Studio 已运行 (默认 http://localhost:1234)');
  console.log('💡 按 Ctrl+C 停止服务');
  console.log('='.repeat(60) + '\n');
});

server.on('error', (err) => {
  if (err.code === 'EADDRINUSE') {
    console.error(`❌ Port ${PORT} is already in use`);
    console.error('   Try: PORT=3000 node serve.js');
  } else {
    console.error('❌ Server error:', err);
  }
  process.exit(1);
});

File: serve.py
#!/usr/bin/env python3
"""
Simple HTTP server for LM Studio Mobile App
Supports single-page app routing for Expo web export
"""

import http.server
import socketserver
import os
import sys
from pathlib import Path

PORT = int(os.environ.get("PORT", 8000))


def find_serve_dir():
    """自动检测静态文件目录。

    - 源码目录：使用 ./dist（如果有构建产物）
    - gh-pages 分支克隆目录：使用当前目录（文件直接在根目录）
    """
    script_dir = Path(__file__).parent

    # 优先使用 dist 子目录
    dist_path = script_dir / "dist"
    if dist_path.is_dir():
        if any(p.suffix == ".html" for p in dist_path.iterdir()):
            return dist_path

    # 检查当前目录是否已经是构建产物（例如 gh-pages 分支）
    current_files = list(script_dir.iterdir())
    has_html = any(p.suffix == ".html" for p in current_files)
    has_expo = (script_dir / "_expo").is_dir()
    if has_html and has_expo:
        return script_dir

    return dist_path


DIST_DIR = find_serve_dir()

class SinglePageAppHandler(http.server.SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=str(DIST_DIR), **kwargs)

    def do_GET(self):
        # Handle routing for single-page app
        requested_path = Path(self.path)

        # If it's a route without extension, try to serve the HTML file
        if not requested_path.suffix and self.path != '/':
            # Remove leading slash and add .html
            file_path = DIST_DIR / f"{requested_path.name}.html"
            if file_path.exists():
                self.path = f"{requested_path.name}.html"

        return super().do_GET()

    def end_headers(self):
        # Add headers to prevent caching during development
        self.send_header('Cache-Control', 'no-store, no-cache, must-revalidate, max-age=0')
        self.send_header('Pragma', 'no-cache')
        self.send_header('Expires', '0')
        return super().end_headers()

    def log_message(self, format, *args):
        # Custom logging
        print(f"[{self.log_date_time_string()}] {format % args}")


def start_server():
    """启动 HTTP 服务器"""
    if not DIST_DIR.exists():
        print(f"❌ 错误：找不到静态文件目录 {DIST_DIR}")
        print("")
        print("请确认以下情况之一：")
        print("  1. 源码目录：运行 npx expo export --platform web --output-dir ./dist")
        print("  2. gh-pages 分支：直接在克隆目录运行 python serve.py")
        sys.exit(1)

    html_files = [p for p in DIST_DIR.iterdir() if p.suffix == ".html"]
    if not html_files:
        print(f"❌ 错误：{DIST_DIR} 中没有找到 HTML 文件")
        print("请先运行 npx expo export --platform web --output-dir ./dist")
        sys.exit(1)

    handler = SinglePageAppHandler

    with socketserver.TCPServer(("", PORT), handler) as httpd:
        print("=" * 60)
        print("🚀 LM Studio Mobile App - Web Server")
        print("=" * 60)
        print(f"\n📂 静态目录: {DIST_DIR}")
        print(f"✅ 服务运行中: http://localhost:{PORT}\n")
        print("🎯 在浏览器打开:")
        print(f"   http://localhost:{PORT}")
        print(f"   http://127.0.0.1:{PORT}\n")
        print("⚠️  请确保 LM Studio 已运行 (默认 http://localhost:1234)")
        print("💡 按 Ctrl+C 停止服务")
        print("=" * 60 + "\n")

        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            print("\n\n🛑 服务已停止")


if __name__ == "__main__":
    start_server()
